add example with python parameters file (#1799)
iterative/dvc.org: content/docs/command-reference/params/index.md

          python train.py
```

> ## Examples: Parameters in Python file

Reviewer: ```suggestion
## Examples: Python parameters file
```

> Сreate a parameters file in Python format, `params.py`:

Reviewer: ```suggestion
Consider this parameters file in Python format, named `params.py`:
```

```python
IS_BOOL: bool = True
CONST = 5

class TrainConfig:
    EPOCHS = 70

    def __init__(self):
        self.layers = 9

class TestConfig:
    TEST_DIR = "path"
    METRICS = ["metric"]
```

The following [stage](/doc/command-reference/run) depends on params

> `IS_BOOL`, `CONST`, `layers`, and `EPOCHS`:

Reviewer: ```suggestion
`IS_BOOL`, `CONST`, as well as `TrainConfig`'s `EPOCHS` and `layers`:
```

> [Stage](/doc/command-reference/run) that depends on params `IS_BOOL`, `CONST`,
> `layers`, and `EPOCHS` from the params file above:

Reviewer: ```suggestion
The following [stage](/doc/command-reference/run) depends on params
`IS_BOOL`, `CONST`, `layers`, and `EPOCHS`:
```

> `IS_BOOL`, `CONST`, as well as `TrainConfig`'s `EPOCHS` and `layers`:

Reviewer 2: @aandrusenko qq - if we depend on the instance variable (`self.layers`), when do we extract value? What is the semantics here?

Author: From a class, you can extract either class constants or `self` variables defined in` __init__`

Reviewer 2: Yep, thanks! But what about the semantics? What if I have

```
self.layers = 10
self.layers = 12
```

or if I have

```
self.layers = 3+2
```

what are the expectations and limitations here?

Author: In the first case, `TrainConfig.layers` will be `12`. In the second, `TrainConfig.layers` will not be found because the complex expression

Reviewer 2: let's make it part of the docs and/or example? @aandrusenko it's totally fine if you don't have enough capacity, let us know and we'll take it over, but I think we should specify some basic semantics for this. Modifying example (add a few lines with Python comments + some text ?) should be enough, I guess.

Author: @shcheklein, I added more comments

```dvc
$ dvc run -n train -d users.csv -o model.pkl \
          -p params.py:IS_BOOL,CONST,TrainConfig.EPOCHS,TrainConfig.layers \
          python train.py
```

> Resulting `dvc.yaml` file:

Reviewer: ```suggestion
Resulting `dvc.yaml` file (notice the `params` field):
```

> You can find that each parameter and it's value were saved to `dvc.yaml`:

Reviewer: ```suggestion
Resulting `dvc.yaml` file:
```

```yaml
stages:
  train:
    cmd: python train.py
    deps:
      - users.csv
    params:
      - IS_BOOL
      - CONST
      - TrainConfig.EPOCHS
      - TrainConfig.layers
    outs:
      - model.pkl

> ```

Reviewer: Should we also/alternatively show the stage in dvc.lock though? Including the param values.

Reviewer: Or at least make a note (md block quote) about that, perhaps.

Author: Added

Reviewer: Thanks @aandrusenko

Alternatively, the entire `TestConfig` group can be referenced, instead of the

> parameters in it:

Reviewer: What other data structures are supported as groups other than classes @aandrusenko @efiop ? Thanks

Author: Only classes and dicts

> Alternatively, the entire group of parameters `TestConfig` can be referenced:
> 
> ```dvc
> $ dvc run -n train -d users.csv -o model.pkl \
>           -p params.py:TestConfig \
>           python train.py
> ```

Reviewer: ```suggestion
Alternatively, the entire `TestConfig` group can be referenced, instead of the
parameters in it:

```dvc
$ dvc run -n train -d users.csv -o model.pkl \
          -p params.py:IS_BOOL,CONST,TestConfig \
          python train.py
```
```

          python train.py
```

## Examples: Python parameters file

Consider this parameters file in Python format, named `params.py`:

```python

> IS_BOOL: bool = True
> CONST = 5
> 
> # All standard variable types are supported
> FLOAT = 0.001
> STR = 'abc'
> DICT = {
>     "a": 1,
>     "b": 2
> }
> LIST = [1, 2, 3]
> SET = {4, 5, 6}
> TUPLE = (10, 100)
> NONE = None

Reviewer: ```suggestion
# All standard types are supported
INT = 5
FLOAT = 0.001
STR = 'abc'
BOOL = True
DICT = {
    "a": 1,
    "b": 2
}
LIST = [1, 2, 3]
SET = {4, 5, 6}
TUPLE = (10, 100)
NONE = None
```
But not sure we want to list them all. People using this feature should know Python 😋 so maybe truncate the list and add a `# etc.`

Reviewer: Implies changes in const names throughout the doc

> # It is possible to retrieve either class constants
> # or own variables defined in __init__

Reviewer: ```suggestion
# DVC can retrieve class constants and variables defined in __init__
```

class TrainConfig:
    EPOCHS = 70

>     def __init__(self):
>         # TrainConfig.layers param will be 9
>         self.layers = 5
>         self.layers = 9
>         # TrainConfig.foo will NOT be found because the complex expression
>         self.foo = 1 + 2
>         # TrainConfig.bar will NOT be found
>         bar = 1

Reviewer: ```suggestion
        self.layers = 5
        self.layers = 9     # TrainConfig.layers param will be 9
        self.sum = 1 + 2    # Will NOT be found due to the expression
```
(The last one is simply not defined in the class so kind of a Python thing.)

Reviewer: Why can't DVC find vars that need evaluation though? Isn't the whole params file interpreted by Python first @aandrusenko @efiop? (Just curious) Thanks

Author: Nope, parameters are parsed using the ast module, it parses complex expressions into a complex structure that I did not support

Reviewer: I see, thanks.

class TestConfig:
    TEST_DIR = "path"
    METRICS = ["metric"]
```

The following [stage](/doc/command-reference/run) depends on params `IS_BOOL`,
`CONST`, as well as `TrainConfig`'s `EPOCHS` and `layers`:

```dvc
$ dvc run -n train -d users.csv -o model.pkl \
          -p params.py:IS_BOOL,CONST,TrainConfig.EPOCHS,TrainConfig.layers \
          python train.py
```

Resulting `dvc.yaml` and `dvc.lock` files (notice the `params` list):

```yaml
stages:
  train:
    cmd: python train.py
    deps:
      - users.csv
    params:
      - IS_BOOL
      - CONST
      - TrainConfig.EPOCHS
      - TrainConfig.layers
    outs:
      - model.pkl
```

```yaml
train:
  cmd: python train.py
  deps:
    - path: users.csv
      md5: 23be4307b23dcd740763d5fc67993f11
  params:
    CONST: 5
    IS_BOOL: true
    TrainConfig.EPOCHS: 70
    TrainConfig.layers: 9
  outs:
    - path: model.pkl
      md5: 1c06b4756f08203cc496e4061b1e7d67

> ```

Reviewer: @shcheklein I think in these cases just dvc.lock is enough in terms of showing all the relevant pipeline/stage structure (although in theory dvc.yaml is more important). Or do you think it's OK to have both? It's kind of long and repetitive.

Reviewer 2: I think it's fine to have both. It's a minor nuance to my mind.

> Alternatively, the entire `TestConfig` group can be referenced (also a
> dictionary), instead of the parameters in it:

Reviewer: Per https://github.com/iterative/dvc.org/pull/1799#pullrequestreview-497944318:
```suggestion
Alternatively, the entire `TestConfig` group (class) can be referenced
(dictionaries are also supported), instead of the parameters in it:
```
Maybe link class to https://docs.python.org/3.8/library/stdtypes.html#classes-and-class-instances
and dictionary to https://docs.python.org/3.8/library/stdtypes.html#dict

Author: I agree
